sandman/source/logger.h

```diff
@@ -1,79 +1,351 @@
+	{
+		return m_OutputStream;
+	}
+
+	[[nodiscard]] constexpr bool HasScreenEchoEnabled() const
+	{
+		return m_ScreenEcho;
+	}
+
+	// "Lower-level" write function.
+	// This simply writes data to the internal buffer,
+	// then flushes the data to the output destinations.
+	//
+	// The internal buffer should be empty after every call to this function.
+	template <typename T, typename... ParamsT>
+	[[gnu::always_inline]] inline void Write(T&& firstArg, ParamsT&&... args);
+
+	static constexpr char kFormatEscapeIndicator{ '%' }, kFormatInterpolationIndicator{ '$' };
+
+	static constexpr std::string_view kMissingFormatValue = "`null`";
+
+	// "Lower-level" format write function.
+	// 
+	// Base case for template recursion. 
+	void FormatWrite(std::string_view const formatString);
+
+	// "Lower-level" format write function.
+	// 
+	// Interpolates arguments into the format string
+	// as it writes all characters in the format string to
+	// the buffer.
+	// 
+	// `%$` interpolates an argument into the format string.
+	// `%%` writes a literal percent sign `%` character.
+	//
+	// If interpolation is denoted with `%$` but there are no more arguments left
+	// to interpolate, `::Logger::kMissingFormatValue` is written instead.
+	// 
+	// The arguments are interpolated into the string in the order
+	// that they are passed into this function from left to right.
+	//
+	// If there are more arguments than there are interpolation indicators `%$`
+	// in the format string, the extra arguments are ignored.
+	template <typename T, typename... ParamsT>
+	void FormatWrite(std::string_view formatString, T&& firstArg, ParamsT&&... args);
+
+	// Object wrapper for formatting.
+	//
+	// Instances of this class can be passed to `Logger::WriteLine`.
+	// If so, the arguments will be interpolated into the
+	// format string and written to the buffer as if
+	// `Logger::FormatWrite` was called with the arguments
+	// that the `Format` instance was constructed with.
+	template <typename... ObjectsT>
+	struct Format
+	{
+		std::tuple<ObjectsT...> m_Objects;
+		std::string_view m_FormatString;
+
+		// Constructor.
+		//
+		// `%$` interpolates an argument into the format string.
+		// `%%` writes a literal percent sign `%` character.
+		//
+		// If interpolation is denoted with `%$` but there are no more arguments
+		// to interpolate, `::Logger::kMissingFormatValue` is written instead.
+		//
+		// The arguments are interpolated into the string in the order
+		// that they are passed into this function from left to right.
+		//
+		// If there are more arguments than there are interpolation indicators `%$`
+		// in the format string, the extra arguments are ignored.
+		template <typename... ParamsT>
+		[[nodiscard]] explicit Format(std::string_view const formatString, ParamsT&&... args)
+			: m_Objects(std::forward<ParamsT>(args)...),
+			  m_FormatString(formatString) {}
+
+		// Convenience method.
+		//
+		// Calls `::Logger::WriteLine(*this)`.

```

Comment:
The leading `::` refers to the global namespace.

I was on the fence as to whether I should keep this function `::Logger::Format<ObjectsT...>::WriteLine` in the code.

It only exists for convenience, but it isn't necessary, and the alternative isn't bad. Let me know if I should remove it, as I am leaning toward removing it.
<br>

```C++
void ExampleFunction() {
	static constexpr char const* kDirection{"north"};

	// Prints "The library is north of here.".
	Logger::WriteLine(Logger::Format("The library is %$ of here.", kDirection));

	/*
		This prints the same string as the previous statement.

		The only thing better about this way is that `Logger::Format` doesn't
		have to be typed twice.

		I don't think this is useful enough actually,
		and its confusing because `::Logger::WriteFormattedLine` exists which
		can be confused with `::Logger::Format`.
	*/
	Logger::Format("The library is %$ of here.", kDirection).WriteLine();
}
```